pilot/pkg/networking/core/v1alpha3/listener_waypoint.go

```diff
@@ -94,97 +95,42 @@ func (lb *ListenerBuilder) serviceForHostname(name host.Name) *model.Service {
 
 func (lb *ListenerBuilder) buildWaypointInbound() []*listener.Listener {
 	listeners := []*listener.Listener{}
-	// We create 4 listeners:
-	// 1. Our top level terminating CONNECT listener, `inbound TERMINATE`. This has a route per destination and decapsulates the CONNECT,
-	//    forwarding to the VIP or Pod internal listener.
-	// 2. (many) VIP listeners, `inbound-vip||hostname|port`. This will apply service policies. For typical case (not redirecting to external service),
-	//    this will end up forwarding to a cluster for the same VIP, which will have endpoints for each Pod internal listener
-	// 3. (many) Pod listener, `inbound-pod||podip|port`. This is one per inbound pod. Will go through HCM if needed, in order to apply L7 policies (authz)
-	//    Note: we need both a pod listener and a VIP listener since we need to apply policies at different levels (routing vs authz).
-	// 4. Our final CONNECT listener, originating the tunnel
+	// We create 3 listeners:
+	// 1. Decapsulation CONNECT listener, `inbound TERMINATE`.
+	// 2. IP dispatch listener, handling both VIPs and direct pod IPs.
```

Comment:
Interesting, I hadn't considered doing this outside the decap. Are there substantial differences between the two approaches?

I guess option 1: 1 big route table, N listeners
This option: 1 tiny route, 1 big listiner with N filter chains

So this seems like it may be a bit better in that regard, but wasn't sure if there were other benefits I missed?

Reply:
We can implement a matcher based on PRT, for example, instead of a fixed list of IPs. That would be harder with RDS. In general, all filter chains are very similar so it makes sense to share a listener.

Reply:
Oh, and it also fits better the model that HBONE listener is strictly about decap so that it can potentially be replaced with ztunnel. Trying to do policy there seems like it'd cause issues later.

The only downside is that we won't do telemetry at HBONE I think. Which is fine since we need to parse HTTP inside anyways.

Reply:
You could link ztunnel rust code into envoy and bridge them with the user space socket (which is C mostly). Not that I want to do it, but we shouldn't close the door for that.